sphinx/source/docs/dev_guide/bokehjs.rst

```diff
@@ -121,98 +121,226 @@ Use ``node make help`` to list all available commands.
+output to each test's dedicated baseline files.
+
+Each test in ``test:integration`` consists of two types of baseline comparison:
+
+Textual baseline comparison
+  For each test, the testing framework compares the pixel location of certain elements
+  in the visual output to pixel locations in a baseline file. The pixel locations are a
+  set of bounding boxes. They are stored as plain text in each test's ``.blf`` files.
+
+Visual baseline comparison
+  For each test, the testing suite does a pixel-by-pixel comparison of a screenshot
+  and a baseline image. These baseline images are stored as ``.png`` files. In contrast
+  to textual baseline comparisons, visual baseline comparisons are platform-dependent.
+  Even minor differences in font rendering, for example, will make the pixel-by-pixel
+  comparison fail.
+
+The visual baseline comparison tests are located in the
+:bokeh-tree:`bokehjs/test/integration/` folder and sub-folders.
+:ref:`Bokeh's CI <devguide_testing_ci>` runs these tests on Linux, macOS, and Windows
+environments. The baseline files for each environment are located in the
+:bokeh-tree:`bokehjs/test/baselines/` folder.
+
+Follow these steps to write new visual tests or update existing tests:
+
+1. Create or update visual testing scripts:
+    To write a visual test for BokehJS' testing suite, start by importing the
+    ``display()`` and ``fig()`` functions from the testing framework's ``_util`` module:
+
+    .. code-block:: TypeScript
+
+        import {display, fig} from "../_util"
+
+    The ``display()`` function in ``_util`` is similar to BokehJS' standard ``show()``
+    function: ``display()`` accepts the same arguments as ``show()`` but also captures
+    the visual output for comparison.
+
+    The ``fig()`` function in ``_util`` is similar to BokehJS' standard ``figure()``
+    function: ``fig()`` accepts the same arguments as ``figure()``. However, to keep
+    visual tests as efficient as possible, you should only use ``width`` and ``height``,
+    if possible.
+
+    Keep the width and height of your testing plot as small as possible while still
+    being able to see details the details you want to test with the naked eye. Have as
+    few elements as possible on your plot.
+
+    Follow this general pattern for visual tests:
+
+    .. code-block:: TypeScript
+
+      describe("Object", () => {
+        it("should show certain behavior", async () => {
+          const p = fig([width, height])
+
+          ...
+
+          await display(p)
+        })
+
+2. Run tests locally:
+    Run ``node make tests`` to test your changes on your system. To only run integration
+    tests, use ``node make test:integration``.
+
+    If you want to only run a specific test, use the ``-k`` argument and supply a search
+    string. The search string is case sensitive. The BokehJS testing suite tries to
+    match your search string to the strings defined in the code's ``it()`` and
+    ``should()`` functions. For example:
```

Comment:
```suggestion
    ``describe()`` functions. For example:
```
there is no ``should()`` function, I remember I mentioned it but it was an example of what other framework's keywords